### ISSUE TITLE
decode_mesh error

### ISSUE THREAD

**Issue by orangeCantaloupe66** [uvic-aurora/edgebreaker#2]
I encountered this error during the decompression process, how can I fix it? (There is no problem with the input off file)
![屏幕截图 2024-01-18 192120](https://github.com/uvic-aurora/edgebreaker/assets/119167026/015d399f-8dcc-4a8c-bbc1-fb40849581c2)

**Comment by mdadams** [comment]
It is impossible to say what the problem might be based on the very limited information that you have provided.  Please provide a reproducer for the problem.

**Comment by orangeCantaloupe66** [comment]
Thank you for your reply！ Now I will give the reproducer：
#Issue Title
decompression errors

#Environment
OS: Windows 10
library version: SPL 2.0.8, CGAL 4.14, boost 1.70.0

#Steps to Reproduce
1. I first used the following command to compress 2h.off, this process successfully generated 2h.eb
encode_mesh <2h.off> 2h.eb

2. Then I used the following command to decompress 2h.eb, and these errors appeared.
decode_mesh <2h.eb> 2h_dec.off

ERROR: Decode regular failed 5
ERROR: Decode y coordinate failed
ERROR: Failed to read the compressed triangle mesh!

#Additional Information
The off file I used and error log/screenshot are as follows
Please see attachment for off file（It seems that the .off file cannot be uploaded, so I uploaded the .txt file）
[2h.txt](https://github.com/uvic-aurora/edgebreaker/files/13983752/2h.txt)
![屏幕截图 2024-01-18 192120](https://github.com/uvic-aurora/edgebreaker/assets/119167026/41821965-5bc7-46ed-aaa7-cf1fe05a0338)

**Comment by mdadams** [comment]
Did you read the detailed documentation on the usage of the software?  The input OFF data matters very such, since vertex coordinates are quantized.

Please be sure to read the Chapter 4 of the report:
https://github.com/uvic-aurora/edgebreaker/blob/master/report/yuetang_meng_project_report.pdf
If you do not do this, you may be using the software in a completely unreasonable way.
Quantization error can cause the mesh to behave very strangely, which could result in encoder/decoder failures.

**Comment by orangeCantaloupe66** [comment]
I did quantize the coordinates, and I used the default quantization step in the paper (as shown in the picture below)
![image](https://github.com/uvic-aurora/edgebreaker/assets/119167026/2b94b7d9-547c-4555-ba9c-bd6c86b93b0b)
At the same time, I set the number of bits to encode each coordinate to 16 bits, however the problem still exists
![image](https://github.com/uvic-aurora/edgebreaker/assets/119167026/4f88188e-914f-4a32-bcf3-7efa0287ad47)
what should I do?

**Comment by mdadams** [comment]
Is your mesh still a valid mesh after vertex quantization?  Quantization has the effect of randomly perturbing every vertex location, and any movement of vertices (no matter how small) has the potential to invalidate a mesh (e.g., by introducing illegal intersections).   The encoder/decoder assumes a valid quantized mesh.  You have to ensure the validity of the mesh yourself, as this is not the responsibility of the coder.

**Comment by mdadams** [comment]
@orangeCantaloupe66 This code was written by one of my graduate students a few years ago.  I hadn't used the code myself in some time.  I had some problems running the code  today, as there were latent bugs caused by missing return statements in three assignment operator functions.  I have fixed this.  Can you please try the code with these bugs fixes and see if it works with your mesh dataset?  There is good chance that this might fix your problem, if vertex quantization was not the issue.

**Comment by mdadams** [comment]
@orangeCantaloupe66 Did the bug fixes that I mentioned above fix your problems?

**Comment by orangeCantaloupe66** [comment]
I'm sorry that I couldn't reply to you in time due to some personal reasons.
I added these return statements, however the problem persists. Have you tried the input I used? Will you also encounter these problems?
Thank you again for your patient reply. Attached is the off file I used.
[2h.txt](https://github.com/uvic-aurora/edgebreaker/files/14445305/2h.txt)

**Comment by mdadams** [comment]
@orangeCantaloupe66:

 I added some very trivial test cases (as basic sanity checks) to the mesh coder software.  If you build the "check" target with CMake, CMake will run these tests.  Could you try this and see if these test cases pass?  I am wondering if there is something about your platform (Windows) that is causing the encoder/decoder to fail for all inputs, not just the file that you provided.  The "check" target's test cases will try to code all of the meshes in the data directory.

I am able to run the encoder and decoder successfully for the OFF file that you provided on Fedora Linux and Ubuntu Linux.  I suspect that the bug is either related Windows/MSVC, quantization, or integer overflow.  If you are on a machine with 64 bit integers, overflow is less likely to be an issue, I think.  I am not sure about 32 bits, however.  Windows might set "int" to 32 bits (or even worse 16 bits?).  If you have Windows Subsystem for Linux (WSL), I would suggest trying the code there as well.  It would be interesting to see if it works in that environment.  Also, I think that the Windows shell might open files for output redirection in text mode.  It is critical that binary mode be used.  If text mode is used for redirection, all binary files generated by the encoder will be corrupted.

**Comment by mdadams** [comment]
@orangeCantaloupe66: I also noticed a few places in the code where the original author did not check for I/O errors in the decoder.  I added a few more checks, but some are still missing.  At least, a read error in the decoder is statistically more likely to be noticed now.  As best I can tell, however, your issue is probably not caused by this type of bug. since you are getting an "ERROR" message (not a silent failure).
